### PR TITLE
Remove String IsASCII flag from the object header

### PR DESCRIPTION
**Pull request by EgorBo**
Not sure if this PR should be merged but at least it's a nice diff to understand how this feature works.
I even made a schema (click to zoom in):

image

This control bit in the object header signals if the string is checked that it doesn't contain any high char (>0x80), see this comment from syncblk.cpp:
```csharp
    //the highest 2 bits have reloaded meaning
    //for string objects:	
    //         BIT_SBLK_STRING_HAS_NO_HIGH_CHARS   0x80000000	
    //         BIT_SBLK_STRING_HIGH_CHARS_KNOWN    0x40000000	
    //         BIT_SBLK_STRING_HAS_SPECIAL_SORT    0xC0000000	
    //for other objects:	
    //         BIT_SBLK_FINALIZER_RUN              0x40000000
```
Currently this flag works only on .Unix systems and can be checked via `String.IsFastSort` internal call used by internal IndexOf for the invariant cultures and for example `String.Replace(string,string,Invariant)` uses it.

so when we do:
```csharp
string str2 = str.Replace("a", "b", StringComparison.InvariantCulture);
```
if we call `Replace` for the first time it scans `str` for the "high chars" and sets the bit to 0x80 (no high chars) or to 0x40 (has high chars) - see [IsFastSort](https://github.com/dotnet/coreclr/blob/0e15f47d8e1c2be4919c49fe35f9d50ede66a057/src/classlibnative/bcltype/stringnative.cpp#L44-L46) (IsFastSort is called from [here](https://github.com/dotnet/coreclr/blob/de68c9ddd18f863fd67098ab28d19206c9c66627/src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs#L249-L258))
so next time we call `Replace` for this string the call will be a little bit faster (since `IsFastSort` will be O(1) instead of O(N)).

BTW, if we want to keep this flag why string literals don't have it set by default?
BTW2, if a string is checked that it doesn't contain any high char and has this flag set to 0x80 and we then modify the string via e.g.:
```
fixed (char* c = str) *c = 'Й';
```
The `IsAscii` flag will remain and the comparison logic will be broken (I know, modifying strings this way is not a good idea but anyway).

cc: @jkotas @filipnavara

### CODE REVIEW

**Comment by danmosemsft** [comment]
@egorbo what is the motivation for this change?

**Comment by EgorBo** [comment]
@danmosemsft https://github.com/dotnet/corefx/pull/37358#discussion_r280611000 and https://github.com/dotnet/coreclr/pull/24373 so while this change can make some benchmarks a bit slower (e.g. if they repeatedly call Replace for the same string over and over - in the real world strings tend to be more unique) it also makes code more portable (as you can see I removed #if CORECLR from the code that is "shared" with other runtimes including Mono).

**Comment by EgorBo** [comment]
Also as I said I am not sure it should be removed at all, I did it mostly as an investigation 🙂   and if we want it to stay I wonder why we don't use it on Windows - on Windows it uses pinvoke into kernel32 FastStringOrdinal (https://docs.microsoft.com/en-us/windows/desktop/api/libloaderapi/nf-libloaderapi-findstringordinal) so currently this logic is useless on Windows if I am not mistaken

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs [hunk 1]
```diff
@@ -245,8 +245,7 @@ internal unsafe int IndexOfCore(string source, string target, int startIndex, in
             Debug.Assert((options & CompareOptions.OrdinalIgnoreCase) == 0);
             Debug.Assert((options & CompareOptions.Ordinal) == 0);
 
-#if CORECLR
-            if (_isAsciiEqualityOrdinal && CanUseAsciiOrdinalForOptions(options) && source.IsFastSort() && target.IsFastSort())
+            if (_isAsciiEqualityOrdinal && CanUseAsciiOrdinalForOptions(options) && source.IsAscii() && target.IsAscii())
```
Comment (EgorBo) [inline src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: I guess this code can be optimized - we can omit IsAscii and just run IndexOf and figure out if it's not "IsAscii" on the fly.

**Comment by filipnavara** [comment]
I'm mostly worried about removing the IndexOf optimization resulting in performance regressions.

If the flag was to be kept then there's a lot of places where it can be propagated in various operations (eg. Substring of FastSort string is guaranteed to be FastSort).

**Comment by jkotas** [comment]
> this change can make some benchmarks a bit slower

And make some benchmarks faster. This change needs perf numbers.

> I guess this code can be optimized - we can omit IsAscii and just run IndexOf and figure out if it's not "IsAscii" on the fly.

Correct. Whether or not this is worth doing should be supported by perf measurements. This optimization was used when removing use of this flag in other places, but not all of them.

> there's a lot of places where it can be propagated in various operations

The culture specific operations are not used that frequently, and they are a lot slower than a ordinal operations by design. I do not think it is a good idea to spread peanut butter overhead and complexity everywhere for something that is slow already.

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs (hunk 1 above)
Comment (tarekgh) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: Looks Windows implementation don't have this optimization. we can try just remove the whole optimization and compare the perf with Windows. 

checking IsAscii only here is incorrect in Linux as you can run into issues when having characters like hyphens. 

---
In reply to: [281011011](https://github.com/dotnet/coreclr/pull/24400#discussion_r281011011) [](ancestors = 281011011)

**Comment by tarekgh** [comment]
We should try to remove the ASCII case optimization and have the perf numbers. Using IsAscii only is wrong especially when string has some characters like hyphen or single quotes.

**Comment by jkotas** [comment]
Superseded by https://github.com/dotnet/coreclr/pull/26759

**Pull request closed without merging**